### ISSUE TITLE
jitter option still has problems with systemd reboot

### ISSUE THREAD

**Issue by nicola-lunghi** [nhorman/rng-tools#67]
I tested the current master and even with patch

https://github.com/nhorman/rng-tools/commit/62fbff0ac807e07f4f21f58c0f15bfe5e9c8b55e

I still have problem on reboot with systemd

```
systemd-shutdown[1]: Sending SIGTERM to remaining processes...
systemd-journald[442]: Received SIGTERM from PID 1 (systemd-shutdow).
systemd-shutdown[1]: Sending SIGKILL to remaining processes...
```
after this it hangs until the watchdog kicks in then

```
systemd-shutdown[1]: Sending SIGKILL to PID 437 (rngd).
systemd-shutdown[1]: Hardware watchdog 'imx2+ watchdog', version 0
systemd-shutdown[1]: Unmounting file systems.
systemd-shutdown[1]: All filesystems unmounted.
systemd-shutdown[1]: Deactivating swaps.
systemd-shutdown[1]: All swaps deactivated.
systemd-shutdown[1]: Detaching loop devices.
systemd-shutdown[1]: All loop devices detached.
systemd-shutdown[1]: Detaching DM devices.
ci_hdrc ci_hdrc.2: remove, state 1
usb usb3: USB disconnect, device number 1
usb 3-1: USB disconnect, device number 2
usb 3-1.2: USB disconnect, device number 3
ftdi_sio ttyUSB0: FTDI USB Serial Device converter now disconnected from ttyUSB0
ftdi_sio 3-1.2:1.0: device disconnected
ci_hdrc ci_hdrc.2: USB bus 3 deregistered
imx2-wdt 30280000.wdog: Device shutdown: Expect reboot!
ci_hdrc ci_hdrc.1: remove, state 4
usb usb2: USB disconnect, device number 1
ci_hdrc ci_hdrc.1: USB bus 2 deregistered
ci_hdrc ci_hdrc.0: remove, state 4
usb usb1: USB disconnect, device number 1
ci_hdrc ci_hdrc.0: USB bus 1 deregistered
reboot: Restarting system
```
can you help me debug it?

If I start the service with jitter disabled
```
-x jitter
````
It reboots correctly

**Comment by nhorman** [comment]
run rngd -f -d  instead of running it via systemd while the system is up, then kill it with a SIGTERM from another terminal and send me the output of that

**Comment by nicola-lunghi** [comment]
[screenlog.txt](https://github.com/nhorman/rng-tools/files/3262618/screenlog.txt)

**Comment by nicola-lunghi** [comment]
Ok description:

if I kill rngd before
```
jent_read_entropy time on cpu 0 is 1.265280712500e+01 sec
```

-> you have the first case

```
Initializing available sources
Initializing entropy source hwrng
Limiting thread count to 2 active cpus
JITTER starts 2 threads
JITTER thread on cpu 0 wakes up for refill
JITTER thread on cpu 1 wakes up for refill
Terminated
```

Instead if I wait after
```
Pool full at 4096, sleeping!
```
I got
```
Checking on done for thread 0

jent_read_entropy time on cpu 1 is 1.045238600000e+01 sec

Writing to pipe

DONE Writing to pipe with return -1

Error on pipe write: Bad file descriptor

jent_read_entropy time on cpu 0 is 3.505968987500e+01 sec

Writing to pipe

DONE Writing to pipe with return -1

Error on pipe write: Bad file descriptor

Closing thread 0

Checking on done for thread 1

Closing thread 1
```
but before the closing thread  message is very slow

**Comment by nicola-lunghi** [comment]
Also If I execute
```
rngd -d -l
Entropy sources that are available but disabled

1: TPM RNG Device (tpm)
```

and get stuck for a while
then it prints

```

Available and enabled entropy sources:

0: Hardware RNG Device (hwrng)

5: JITTER Entropy generator (jitter)

Checking on done for thread 0

```
but if I  type
```
rngd -d -x jitter -l
```
is very fast
```
Disabling 5: JITTER Entropy generator (jitter)

Entropy sources that are available but disabled

1: TPM RNG Device (tpm)

5: JITTER Entropy generator (jitter)

Available and enabled entropy sources:

0: Hardware RNG Device (hwrng)

```

**Comment by nhorman** [comment]
you don't have to disable jittier, thats where the problem is that we want to debug.

That said, I'm not sure how this is happening, on my system as soon as close the read side of the pipe, all the writers return immediately with the EBADF error, causing them to break out of their loop and exit.

Let me ask you, when you say "very slow" do you remember about how long it took to exit?  Was it approximately 35 seconds?

**Comment by nicola-lunghi** [comment]
```
time rngd -l
Entropy sources that are available but disabled

1: TPM RNG Device (tpm)

Available and enabled entropy sources:

0: Hardware RNG Device (hwrng)

5: JITTER Entropy generator (jitter)

Command terminated by signal 2
real    0m 12.59s
user    0m 21.41s
sys     0m 3.28s
```

**Comment by nhorman** [comment]
ok, I think I know whats happening.  if you look at 
https://github.com/nhorman/rng-tools/issues/67#issuecomment-499581667

You'll see that the time spend generating entropy in the jitterentropy daemon is :
jent_read_entropy time on cpu 0 is 3.505968987500e+01 sec

Thats about 35 seconds,  While sending pthread_kill(SIGINT) to the writer thread will make it break the loop if its blocked on a write, it won't cause it to return early from that generation function, meaning that you have to wait for some part of that 35 seconds to break out of the loop and exit.  Thats not strictly speaking wrong, but it is a bit of an irritant.  I might could try to square that up with some setjmp/longjump magic, but I'm not sure.

**Comment by nicola-lunghi** [comment]
Ok my issue is that rngd get stuck on reboot
Now I have to look at another issue for now I disabled jitter and is working fine

I will try to add more printf and debug again (sadly is quite difficult to add gdb to this system :-( )

**Comment by nhorman** [comment]
https://github.com/nhorman/rng-tools/tree/early-jitter-exit

I've created the new branch above, which uses sigsetjmp/siglongjump to effect an early return from jent_read_entropy, which should allow for earlier exiting from threads that are delayed by the gathering of entropy from the jitter library.  Please build that branch and give it a try with jitterentropy enabled.  If it works for you I'll merge it.

**Comment by nhorman** [comment]
ping, any update?

**Comment by nicola-lunghi** [comment]
Testing it now

**Comment by nicola-lunghi** [comment]
no differences.

**Comment by nicola-lunghi** [comment]
I will try to look at this next week (now I'm stuck with another issue).
if you want you can close it and I will reopen it with more debug information.

**Comment by nhorman** [comment]
ok, saying no differences without additional information doesn't help.  I'm going to add some debug information to this branch.  I'll leave this open and when you have time, please run rngd with -d so debug information is recorded in the logs, and send that in.
